src/interpreter.c

```diff
@@ -166,6 +166,82 @@ static void eval_abstracttype(jl_expr_t *ex, interpreter_state *s)
     if (temp == NULL || !equiv_type(dt, (jl_datatype_t*)jl_unwrap_unionall(temp))) {
         jl_checked_assignment(b, w);
     }
+    dt->incomplete = 0;
+    JL_GC_POP();
+}
+
+static void check_type_completion(jl_datatype_t *bdt, jl_value_t *super, jl_value_t *para)
+{
+    // Check that supertype matches
+    if (!jl_types_equal((jl_value_t*)bdt->super, super))
+        jl_error("Super type does not match forward declaration");
+    // Check that parameters match
+    if (jl_svec_len(para) != jl_svec_len(bdt->parameters)) {
+        jl_error("number of type parameters does not match forward declared type");
+    }
+    for (int i = 0; i < jl_svec_len(para); ++i) {
+        jl_tvar_t *v = (jl_tvar_t*)jl_svecref(para, i);
+        jl_tvar_t *w = (jl_tvar_t*)jl_svecref(bdt->parameters, i);
+        assert(jl_is_typevar(v) && jl_is_typevar(w));
+        if (v->name != w->name)
+            jl_errorf("Name `%s` of type parameter does not match name `%s` from forward declaration.",
+                jl_symbol_name(v->name), jl_symbol_name(w->name));
+        else if (!jl_types_equal(v->lb, w->lb) || !jl_types_equal(v->ub, w->ub)) {
```

Comment:
The `lb` and `ub` cannot be passed to `jl_types_equal` directly (they might themselvers be TypeVar)

Could we pose this whole loop as a simple question to the type-system? `jl_types_equal(bdt->name->wrapper, jl_apply_type(bdt->name->wrapper, jl_svec_data(para), jl_svec_len(para))`